Fix image caching issues when image is replaced by another image with same name (Fix #380)

mjameswh opened the pull request:
What does this implement/fix? Explain your changes
--------------------------------------------------
Avoid some caching issues causing old version of image being displayed if an image file is replaced by another image file with the same name. More specifically:

- In `ImageCache`, check file's last modification before returning a cache entry.
- In `GraphicsUtils`, avoid creating `BitmapImage` from a file URI (which seems to trigger some internal cache... not sure why) by instead loading the file to a `Bitmap`, then convert it to a memory stream, and finally create a `BitmapImage` from that memory stream.

Does this close any currently open issues?
------------------------------------------
- #380 Wrong image displayed after replacing file (image caching issue)

Any other comments?
-------------------
- Fix has been tested locally.
- This fix introduces some extra steps in loading images (image actually gets decoded as a `Bitmap`, then reencoded to memory stream, then decoded again as a `BitmapImage`). This should be insignificant in most cases, but could be problematic with very large images on machines with low memory (this all happens before the image is reduced to a maximum size). This could be avoided by opening a `FileStream` on the image file and creating the `BitmapImage` directly from that stream. However, at this point, I prefered keeping my change more in line with what's done elsewhere in the code. If desired, I could later do some refactoring on `GraphicsUtils`.

AntonyCorbett commented:
@mjameswh Thanks for taking the time to look into this. The current fix prevents us from removing image files using Windows Explorer. i.e. if you display an image it keeps the file handle open and we can't remove it from File Explorer or from within the OnlyM Command Panel.

mjameswh commented:
@AntonyCorbett You are right... I actually tested for that case, but I realize now that this is sensible to garbage collection timing. A new fix coming up in a few minutes...

mjameswh commented:
@AntonyCorbett One thing I was not able to test is the "image has invalid DPI" case... Guess you have some test files you did used for testing that scenario. If there is a problem there, please provide the test file so I can investigate.

AntonyCorbett commented on OnlyM.CoreSys/GraphicsUtils.cs:
```diff
@@ -237,26 +237,22 @@ public static BitmapImage ByteArrayToImage(byte[] imageData)
-            // BitmapCacheOption.OnLoad prevents the source image file remaining
-            // in use when the bitmap is used as an ImageSource.
-            bmp.BeginInit();
-            bmp.CreateOptions = BitmapCreateOptions.IgnoreColorProfile;
```
We must retain this option to deal with photos that have a bad colour profile.

mjameswh replied:
No problem, I'll try to restore this handling somehow. But then, shouldn't we also take similar precautions in thumbnail generation?

Also... I doubt that systematically discarding color profile is desirable. In the past, I have personally seen a case where an image with large area of "bubble gum pink" would turn out "wine red" because ICC profiles were ignored. Granted, that was an extreme case involving CMYK to RGB conversion, but still, even in common scenario, color profiles are generally a good thing... I know however that WPF will throw when processing an image with an invalid color profile.

Would it sound good to you if I change the code to first try to load the image with its embedded color profile, then catch eventual exceptions, and only then, retry loading the image ignoring color profile? Note that I may be taking over my head now, as I am not familiar with .NET nor WPF, so maybe I will find later that doing this is is ridiculously complex, but if that sounds interesting to you, I am willing to give it a try...

AntonyCorbett replied:
Yes, good ideas all round. Thanks for your patience.

AntonyCorbett commented on OnlyM/Services/ImagesCache/ImageCache.cs:
```diff
@@ -15,49 +15,57 @@ internal class ImageCache
         public BitmapSource GetImage(string fullPath)
         {
-            if (!_cache.TryGetValue(fullPath, out var result))
+            if (!File.Exists(fullPath))
```
Could we use `new FileInfo(fullPath)` here for performance? We could remove `File.Exists` and `File.GetLastWriteTimeUtc`

mjameswh replied:
Indeed, that makes sense. I will fix this.

AntonyCorbett commented:
@mjameswh Thanks for your work on this. Ready for me to pull?

mjameswh commented:
> @mjameswh Thanks for your work on this. Ready for me to pull?

No, not yet. Sorry for the confusion, I pushed very late yesterday and should have provided a little more explanations, but ...

So here is an update on what I did:

- I replaced `File.Exists + File.GetLastWriteTimeUtc` sequences by use of `Fileinfo`, as previously discussed.
- I have identified and fixed a second caching issue, similar to #380, where thumbnails would not be refreshed if a file was _replaced_ (that is, suppose already have a file named "test.jpg" in your media folder, and move or copy another file also named "test.jpg" to that folder, and confirm that you want to replace that file). Regarding that matter, do you prefer that I open a distinct ticket for that issue, for reference/documentation?
- I have created a few test images that demonstrate various corner cases, including an image containing an invalid color profile. Turns out that right now, the code doesn't crash on that image. Actually, color profiles seems to be completely ignored... I will revisit that area afterward, but at this point, I think this would be better handled in a distinct contribution. By the way, for reference, the picture of fruits I used for testing is from Hans Braxmeier, licensed as CC0 Universal Public Domain (see on [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Fruit-49740_bananas_apples_tangerines_-_Hans_Braxmeier.jpg)).
- However, `BitmapToBitmapImage` throws on the image with very high DPI (10000x10000 dpi). Strangely, it doesn't throw anymore if I simply comment the call to `isBadDPI()`, and the image appears just fine. I have to investigate this before moving on with the PR.

Thanks brother for your support on this issue, as well as your work on developing this software. Very appreciated.

AntonyCorbett commented:
@mjameswh I have been working on a .NET 5 port and there have been many code changes. Since this PR is still a draft, I have taken the liberty of including fixes in the new branch. Thanks for your help

mjameswh commented:
> @mjameswh I have been working on a .NET 5 port and there have been many code changes. Since this PR is still a draft, I have taken the liberty of including fixes in the new branch. Thanks for your help

@AntonyCorbett I'm glad to hear this. I unfortunately don't have access to a Windows machine at present, making it difficult for me to progress on this. Thanks for your work.
